driver-core/src/main/com/mongodb/internal/connection/ByteBufBsonArray.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import org.bson.BsonArray;
+import org.bson.BsonBinaryReader;
+import org.bson.BsonType;
+import org.bson.BsonValue;
+import org.bson.ByteBuf;
+import org.bson.io.ByteBufferBsonInput;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+import static com.mongodb.internal.connection.ByteBufBsonHelper.readBsonValue;
+
+final class ByteBufBsonArray extends BsonArray {
+    private final ByteBuf byteBuf;
+
+    ByteBufBsonArray(final ByteBuf byteBuf) {
+        this.byteBuf = byteBuf;
+    }
+
+    @Override
+    public Iterator<BsonValue> iterator() {
+        return new ByteBufBsonArrayIterator();
+    }
+
+    @Override
+    public List<BsonValue> getValues() {
+        List<BsonValue> values = new ArrayList<>();
+        for (BsonValue cur: this) {
+            //noinspection UseBulkOperation
+            values.add(cur);
+        }
+        return values;
+    }
+
+    private static final String READ_ONLY_MESSAGE = "This BsonArray instance is read-only";
+
+    @Override
+    public int size() {
+        int size = 0;
```

Comment:
Since the `ByteBufBsonArray` is read-only and its size doesn't change after initialization, we can consider caching the `size` value to avoid additional computations during each `size()` call.